qualtran/bloqs/mod_arithmetic/coset_representation.py

```diff
@@ -0,0 +1,86 @@
+#  Copyright 2025 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from functools import cached_property
+from typing import TYPE_CHECKING
+
+import attrs
+import sympy
+
+from qualtran import Bloq, bloq_example, BloqDocSpec, QUInt, Register, Side, Signature
+from qualtran.bloqs.basic_gates import CNOT, Hadamard
+from qualtran.resource_counting import BloqCountDictT, SympySymbolAllocator
+
+if TYPE_CHECKING:
+    from qualtran.symbolics import SymbolicInt
+
+
+@attrs.frozen
+class InitCosetRepresntation(Bloq):
+    r"""A state initialization of an integer in the coset representation.
+
+    The cost representation of an integer $k$ modulo $N$ with $c_{pad}$ bits is defined as
+    $$
+        \frac{1}{\sqrt{2^{c_{pad}}}}\sum_{j=0}^{2^{c_{pad}}} \ket{jN + k}
+    $$
+
+    This bloq can be built of only clifford gates ... namely $c_{pad}$ `H` gates on the padding
+    qubitsfollowed by `CNOT` gates implementing the reversible operation $jN+k$.
+
+    Args:
+        c_pad: The number of padding bits.
+        k_bitsize: The number of bits used to represent $k$ ($\geq$ the number of bits of $k$ and $N$).
+        k: The value of $k$.
+        mod: The value of $N$.
+
+    Registers:
+        x: A k_bitsize+c_pad register output register containing the initialized state.
+
+    References:
+        - [Shor's algorithm with fewer (pure) qubits](https://arxiv.org/abs/quant-ph/0601097)
+            section 4.
+        - [How to factor 2048 bit RSA integers in 8 hours using 20 million noisy qubits](https://arxiv.org/abs/1905.09749)
+            section 2.4
```

Comment:
reference format: use newline to separate entries. Don't use `-`. Include author and year